### PR TITLE
github: run spread via github actions

### PR DESCRIPTION
**Pull request by zyga**
This patch runs adds a github actions workflow for running spread
tests. I've used a matrix build, one per system, so that logs are
shorter and any error patterns are easier to spot.

Earlier test runs indicate that each system takes about 30 minutes to
complete. The fastest system is ubuntu-14.06-64 at around 28 minutes.
The slowest systems are ubuntu-core-20-64 and ubuntu-20.04-64 (classic)
at over 47 minutes. While the slowness of core-20 setup is well-known
the slowness of classic is surprising.

Tests are powered by a pool of 32 self-hosted runners. The matrix uses
14 systems per pull request. This means we can instantly start up to two
pull requests and process them in about 40 minutes. The extra workers 
are spare capacity in case something is restarted. Self-hosted workers
do not cost anything and do not take any concurrency slots in github
actions.

Our free plan gives us 20 slots that are used to run unit tests. This
means that up to 20 pull requests are instantly, concurrently tested
(unit tests) and then processed one-by-one by the spread worker pool.

Matrix workers have a fail-fast semantics by default, so any run that
fails will cancel all the peers. We could explore some more ideas that
run a canary run (e.g. ubuntu-16.04-64) and only run the remaining
combinations if the canary succeeds.

We could also explore building snapd.deb and re-packaging snapd.snap and
core.snap once, offering those as downloadable items (in each pull
request). This would allow us to remove a good chunk of constant delay
that is otherwise incurred by each test (and fix some build bugs we
currently have in repackaging, e.g. when building on fedora). This might
offset the cost of the pipeline, cutting the total time as well.

Signed-off-by: Zygmunt Krynicki <zygmunt.krynicki@canonical.com>

### CODE REVIEW

File: .github/workflows/sprd.yaml [hunk 1]
```diff
@@ -0,0 +1,37 @@
+name: Spread 
+on:
+  push:
+    branches: [ master ]
+  pull_request:
+    branches: [ master ]
+jobs:
+  GCE:
+    runs-on: [self-hosted]
```
Comment (anonymouse64) [inline .github/workflows/sprd.yaml]: out of curiosity what connects the github action runs from this repo to the self-hosted thing. Is there config on the repo itself, or do you just provide some token/auth data to the self-hosted worker when you start it up? If this setup is security sensitive, we can also chat offline tomorrow too

Comment (zyga) [reply .github/workflows/sprd.yaml]: This is a property of the repository. We have registered self-hosted workers that talk to GitHub and wait for requests. Each worker runs in a rootless container, without ability to run sudo (inside the container). The authentication is indeed based on a token that was obtained when each worker was registered.

Comment (zyga) [reply .github/workflows/sprd.yaml]: This process is explained in detail in https://help.github.com/en/actions/hosting-your-own-runners/about-self-hosted-runners

File: .github/workflows/sprd.yaml [hunk 2]
```diff
@@ -29,3 +29,9 @@ jobs:
       uses: actions/checkout@v2
     - name: Run spread tests
       run: spread -v google:${{ matrix.system }}:tests/...
+    - name: Discard spread workers
+      run: |
+        shopt -s nullglob;
+        for r in .spread-reuse.*.yaml; do
+          spread -discard -reuse-pid="$(echo "$r" | grep -o -E '[0-9]+')";
```
Comment (bboozzoo) [inline .github/workflows/sprd.yaml]:
```suggestion
          r="${r#.spread-reuse.}";
          r="${r%.yaml}";
          spread -discard -reuse-pid="$r";
```

Comment (zyga) [reply .github/workflows/sprd.yaml]: I'll commit this on the next pass.

**Comment by zyga** [comment]
Per @mvo5's request I've flipped this around, running unstable systems through github actions and stable systems through travis.

**Pull request merged**